packages/admin/cms-admin/src/generator/future/generator.ts

```diff
@@ -22,31 +25,42 @@ export type FormFieldConfig<T> = (
     | { type: "staticSelect"; values?: string[] }
     | { type: "asyncSelect"; values?: string[] }
     | { type: "block"; block: BlockReference }
-) & { name: keyof T; label?: string; required?: boolean };
+) & { name: string; label?: string; required?: boolean };
+export type FormFieldConfig<T extends GeneratorEntity> = FormFieldConfigInternal & { name: Leaves<T> | Paths<T> };
 
-export type FormConfig<T extends { __typename?: string }> = {
+export type FormConfigInternal = {
     type: "form";
-    gqlType: T["__typename"];
+    gqlType: string;
     fragmentName?: string;
-    fields: FormFieldConfig<T>[];
     title?: string;
+    fields: FormFieldConfigInternal[];
+};
+export type FormConfig<T extends GeneratorEntity> = FormConfigInternal & {
+    gqlType: T["__typename"];
+    fields: FormFieldConfig<T>[];
 };
 
 export type TabsConfig = { type: "tabs"; tabs: { name: string; content: GeneratorConfig }[] };
 
-export type GridColumnConfig<T> = (
+export type GridColumnConfigInternal = (
     | { type: "text" }
     | { type: "number" }
     | { type: "boolean" }
     | { type: "date" }
     | { type: "dateTime" }
     | { type: "staticSelect"; values?: string[] }
     | { type: "block"; block: BlockReference }
-) & { name: keyof T; headerName?: string; width?: number };
-export type GridConfig<T extends { __typename?: string }> = {
+) & { name: string; headerName?: string; width?: number };
+export type GridColumnConfig<T extends GeneratorEntity> = GridColumnConfigInternal & { name: Leaves<T> | Paths<T> };
```

Comment:
I tried improving this, this is what I came up with.
```
type UsableFieldsInner<T> = "__typename" extends keyof T
    ? {
          [K in keyof T]: "__typename" extends K
              ? never
              : `${Exclude<K, symbol>}${UsableFields<NonNullable<T[K]>> extends never ? "" : `.${UsableFields<NonNullable<T[K]>>}`}`;
      }[keyof T]
    : never;
type UsableFields<T> = UsableFieldsInner<Required<T>>;
```

My idea was to use "Leaves", but stop at nested objects that don't have a __typename (as blocks)

Reply:
@nsams I'm sorry your type also has issue with typing in generator-code. could it work with a different type in `FormConfig<any>` instead of any? I tried `FormConfig<{ __typename?: string; [key: string]: any }>` but it didn't work.

<img width="1101" alt="Screenshot 2024-02-23 at 12 57 18" src="https://github.com/vivid-planet/comet/assets/1324250/f08b4e12-a030-47ec-8272-74330641fbba">
---

for `Leaves<T> | Paths<T>` the issue is like this

<img width="1128" alt="image" src="https://github.com/vivid-planet/comet/assets/1324250/b49fbc79-eb93-4a07-86b5-a0c579fbd8f1">
---

With `UsableFields` an internal-type will be required, but at least there are less options provided for name-autocompletion. should I replace the type definition?

Reply:
can you create a ts playground where the issue is reproducable?

Reply:
> can you create a ts playground where the issue is reproducable?

which of the two?

Reply:
UsableFields

Reply:
playground with reproduced error: [Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAqgzgQwEYBsIDECWEUBM4CSAdkRAE4A8AKgHxQC8UARAPouiREIC2ETUEAB7AIRfFADWEEAHsAZlCoBYAFBR1UAPxQA3qo0GNAbQDSUTEUnT5igLoAuZmw6iefAcNHiT+w3-XapABu5L7+ho4ABgAkOgCiggDGKACuuBAUJgA0UHAg3EgyKDQAvrHwyGhYOPgUAHIyRHUpKCiVGVSmtjR0QiJicFDB5FrM-FEAdOWIqBjYeHD1jc2t7dRdPSWRWwDcYRolRlKyClS2+47DZHsqLrAzVfO1tAz37dULxKSUAEoQAI4pTBkCC4ag9G6qO4AWRAVHA0EYOiGbkcFVmH1qCCIIDoJUhKjkKSIiWAmEaUG4cIgcGA6GJpPJRAAFFSAGoIMiOWHwyAASl0+0wClZIA5ZAmXF4EwsyTSNOZTBEtKYfIFejUB1UJSAA)

Reply:
the new version worked like a charm. I replaced the `Leaves | Nodes`. Now also FormConfig<any> is just only a "ignore any" problem and no further typescript issue.

Reply:
ok, cheered to soon. the new version still has issues. Updated the [Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBA4gjgGwDIQIYDcIGcA8AVAPigF4oAiAfQtEgDtUBbCMqCAD2AloBMsoBrCCAD2AMyh4oAfgBQAbxlQlUANoBpKAEtaAoWIlRUfDe048+lauC6Nm0qLQiYATlABcUNQF0AtFI-wyGiYuHjqXgQyAL7uDk4QzgDcMjI00ABimhAI3ABytqFEpApK6lo6giLieF4eAAYAJHIAomwAxggArtwQOGoANFBYIAwARsIIBFFNmdl5BfjhRKZcvHEu9mQs9QB0M1k5+UyhS1F1Z8lRKpX6Ncmp1lAAqlioowgQszmFJFBf88ccIEUBhsPgCJEZAB6ABUUCoCMRiIk2GAUAAwkZsPCkbiKFAYVCHpAoK1GGAPoE8I9ioplFQ0vQmB4yGSGBTmP06UpRMJhB4sMBnNoAObJZRQUaoZz+IZC0Xi5SOQUQbgeEoSnGM2wstkc3Ko1VkLmayWoABeAvltDF3OUUvNACZZYLhTbFUoonblZxuI71XalGxg1a3baJV6vcToHrKYhqZAAOqaYAACwA8mBgJphPQEAaVdxfhqlAzrEyILq2OSPsa7bz+XKwx6zTLQwrvYbuLKS-SrHQdeRYxAC7666aHe33YGzU6Xdbw8ovVCoRKfaqAyvTWWB8yh9X2R9R0auVvNZOmx2zxKHc6p7br1AYgAfBydBAIFvrv0B03Btj3i2kYpGkUCcIKACMvwvG8Hz-Lgw5UtYBDJKB4HAI60GvO8nwHLwOCoLQIAodGYGogAzL8wLBGCiHxshqGPOhAAsWGwbhcwIQeHJIUmKYZlmOZ5se3AkTImjssIzhoqBciknAnSoAggytJAbRojEojOMIDBQAA5AAAmkPhtKmSkfDa2BQp02YIFgekgY8bRYnwpAqHSqkQOpODNApSk4GQDYsK+ZBSs4wXkN+OwOhFZBRbesXfo6xrPNhcF4VxNYQLxEAQgQJqkmwanAD5fkIDgrqioMME4fBBFEXlBWed5vmKeVgV8rFYWJV20UWj1hZ9U6A2+sl1VpRx3w+dxcYIAmEDJmmmbZrmSkiY1MheDIQA)

Reply:
:(

and you are not able to fix this yourself?

Reply:
I can try

Reply:
I managed to solve the optional nested objects problem. [Playground](https://www.typescriptlang.org/play?#code/FAFwngDgpgBA4gRwDYBkoEMBuUDOAeAFQD4YBeGAmKADxCgDsATHGdesGYGGAfhgCIA+oPDR66ALZR+VWg2YwA1lDAB7AGYVO3bjy47uAb30HuAbQDSMAJb0lKjVvQsrNOkxZCRkBpOm8YeihsACcYAC4YCwBdAFoeSMRUDGx8AktoohMdAF9syKDQ-QLgqBCAbmBQHxgAMWsoJEYAOT80knJjcytbezVNAmjIgAMAEkMAUWoAYyQAV0YoPAsAGhgcMAkAI1UkIhzx+saWtsIMkjd5FkKygP4ZEYA6Q4am1qk085zh78qcs2U-Qo0Uq1WgMAAqjh0FskFAjk12mQ6q8Th88Ek0FhcIQiFlgAB6ABUMGEZPJ5IouBAMAAws5cKSKczBDAiQSwbAppIIHCkgQap1ssJRL4pJF+NyJLzpCtsupVKpIjgQCFbABzSoGLboEIJdaqjVanRBFVQRiRLqmEU+cTigRSmXNanm-hy0wwHUAL2Vhvomuy3G9ACZ9Sq1f7jdw8gZTXRGMHLYGYNRU76IwGDHk8pyYI6+cgBdAAOrWEAACwA8hAQNZVOIkM6zYxkVbuDaxH4JfnZfLFemjdkdXqB5HsnHzfq2zoO2KoN3qDy4U3427k97R5nTCGw36t7GXYwACJQKAQJMe9veTv2yWL6XLw-BteXz3oH0GjNR7fv0Ob7-RsmE7HqeEDVrW9boEgU7Jleop2vODr3k6T4vpeG6foOr47v+yYxrkMAAD6BHMSBIOOT4XqYqbULhWbADmoowHQKoAIzIlCMJwgizB4D2-I+EQlRMSxIDBhx0KwvCqL4GwYBCbmokAMzIpiKQ4vxhaCcJNSiQALBJXHScc+CaUgRZQKWFbgXWDYruaCkidSACsqnIFiqR8chBbmT4VlVjWtlQfZjAKcA1jSqoIQ0kxhh5ggcxQWsUzQNMNI5DA6ghKoEgwAA5AAAqKsTTOWUFwv6uAEnMtZIDgeVVEx0wMiw5BmPoKVQGlfEJVBeD8AqqgyMR-DDsNAjAY83rjfwk0hjNwHPmsnFSTxpneVAAnQHiRDunm1CpSAPWJUgeDhhqy2SdxMl4HJO17Z13UTL1p3ZANirDW9Y1EW9c3vuNv2HlNv4AwYs1A8BJ5nsDXqgzo4PNo8kOgTDz5Edo8OTcjZ42ZBSAw3D3AI-GSOHlDYGBXjqOfWDi38Nkl1GWtXlLptWklmWAUQXZh73cA0TAEAA)

Reply:
Seems to work :+1:

(but I don't understand it)

Reply:
@nsams as mentioned yesterday the UsableFields-Type has issues with asyncSelect requiring to reference a graphql object. Do you have an idea what to change to add this type of keys? I updated the [playground](https://www.typescriptlang.org/play?#code/FAFwngDgpgBA4gRwDYBkoEMBuUDOAeAFQD4YBeGAmKADxCgDsATHGdesGYGGAfhgCIA+oPDR66ALZR+VWg2YwA1lDAB7AGYVO3bjy47uAb30HuAbQDSMAJb0lKjVvQsrNOkxZCRkBpOm8YeihsACcYAC4YCwBdAFoeSMRUDGx8AktoohMdAF9syKDQ-QLgqBCAbmBQHxgAMWsoJEYAOT80knJjcytbezVNAmjIgAMAEkMAUWoAYyQAV0YoPAsAGhgcMAkAI1UkIhzx+saWtsIMkjd5FkKygP4ZEYA6Q4am1qk085zh78qcs2U-Qo0Uq1WgMAAqjh0FskFAjk12mQ6q8Th88Ek0FhcIQiFlgAB6ABUMGEZPJ5IouBAMAAws5cKSKczBDAiQSwbAppIIHCkgQap1ssJRL4pJF+NyJLzpCtsupVKpIjgQCFbABzSoGLboEIJdaqjVanRBFVQRiRLqmEU+cTigRSmXNanm-hy0wwHUAL2Vhvomuy3G9ACZ9Sq1f7jdw8gZTXRGMHLYGYNRU76IwGDHk8pyYI6+cgBdAAOrWEAACwA8hAQNZVOIkM6zYxkVbuDaxH4JfnZfLFemjdkdXqB5HsnHzfq2zoO2KoN3qDy4U3427k97R5nTCGw36t7GXYwACJQKAQJMe9veTv2yWL6XLw-BteXz3oH0GjNR7fv0Ob7-RsmE7HqeEDVrW9boEgU7Jleop2vODr3k6T4vpeG6foOr47v+yYxrkMAAD6BHMSBIOOT4XqYqbULhWbADmoowHQKoAIzIlCMJwgizB4D2-I+EQlRMSxIDBhx0KwvCqL4GwYBCbmokAMzIpiKQ4vxhaCcJNSiQALBJXHScc+CaUgRZQKWFbgXWDYruaCkidSACsqnIFiqR8chBbmT4VlVjWtlQfZjAKcA1jSqoIQ0kxhh5ggcxQWsUzQNMNI5DA6ghKoEgwAA5AAAqKsTTOWUFwv6uAEnMtZIDgeVVEx0wMiw5BmPoKVQGlfEJVBeD8AqqgyMR-DDsNAjAeN-DAY83pTTNIbzahaycVJPGmd5UACdAeJEO6ebUKlIA9YlSB4OGGorZJ3EyXgcm7ftnXdRMvVndkA2KsN71jUR72Tb9BjTYes3vuNf3A4tRHaDoQPNo8wEnmeX2AzNCOgSDXpgyjwNo2eGPPgDMOo4eiNgYFkFIMjRM4yToE2RTGNY9TcO42TEENvjVPcLD8bPtkV1GetXlLltWklmWAXs8Fh4PcA0TAEAA) with the new requirements.